aten/src/ATen/native/mps/operations/BinaryKernel.mm

```diff
@@ -1,10 +1,387 @@
+  ScalarType common_dtype = iter.common_dtype();
+  if (isIntegralType(common_dtype, true)) {
+    // integer inputs must be cast to float, if output is float
+    if (isFloatingType(outputDataType)) {
+      common_dtype = outputDataType;
+    // in boolean comparison ops with signed vs. unsigned integers, we always cast to the unsigned type
+    } else if (outputDataType == ScalarType::Bool &&
+              (inputDataType == ScalarType::Byte ||
+                otherDataType == ScalarType::Byte)) {
+      common_dtype = ScalarType::Byte;
+    }
+  }
+
+  if (iter.tensor(1).numel() == 1 && iter.tensor(2).numel() == 1) {
+    all_scalar = true;
+  }
+  else if (iter.tensor(1).numel() == 1 || iter.tensor(2).numel() == 1) {
+    scalar_pos = iter.tensor(1).numel() == 1 ? 1 : 2;
+    tensor_pos = scalar_pos == 1 ? 2 : 1;
+    inputTensor = iter.tensor(tensor_pos);
+    otherTensor = iter.tensor(scalar_pos);
+  } else {
+    inputTensor = iter.tensor(1);
+    otherTensor = iter.tensor(2);
+  }
+
+  if (inputTensor.numel() == 0 || otherTensor.numel() == 0) {
+    return;
+  }
+
+  if (inputTensor.scalar_type() == kDouble) {
+    inputTensor = inputTensor.to(iter.common_dtype());
+  }
+  if (otherTensor.scalar_type() == kDouble) {
+    otherTensor = otherTensor.to(iter.common_dtype());
+  }
+
+  bool allContiguous = false;
+  if (inputTensor.is_contiguous() && otherTensor.is_contiguous() && outputTensor.is_contiguous()) {
+    allContiguous = true;
+  }
+
+  MPSStream* mpsStream = getCurrentMPSStream();
+  id<MTLDevice> device = MPSDevice::getInstance()->device();
+
+  id<MTLBuffer> inputBuffer = mps::getMTLBufferStorage(inputTensor);
+  id<MTLBuffer> otherBuffer = mps::getMTLBufferStorage(otherTensor);
+  id<MTLBuffer> outputBuffer = mps::getMTLBufferStorage(outputTensor);
+  uint32_t inputTensorStorage = inputTensor.storage_offset() * inputTensor.element_size();
+  uint32_t otherTensorStorage = otherTensor.storage_offset() * otherTensor.element_size();
+  mps::MPSScalar scalar;
+  if (all_scalar) {
+    type = BinaryKernelType::Scalar;
+    if (iter.is_cpu_scalar(1)) {
+      scalar = mps::getMPSScalar(inputTensor.item(), inputTensor.scalar_type());
+      inputBuffer = (id<MTLBuffer>)getIMPSAllocator()->allocScalarBufferWithValue(&scalar.value, scalar.size).get();
+      inputTensorStorage = 0;
+    }
+    if (iter.is_cpu_scalar(2)) {
+      scalar = mps::getMPSScalar(otherTensor.item(), otherTensor.scalar_type());
+      otherBuffer = (id<MTLBuffer>)getIMPSAllocator()->allocScalarBufferWithValue(&scalar.value, scalar.size).get();
+      otherTensorStorage = 0;
+    }
+  } else if (scalar_pos) {
+    type = allContiguous ? BinaryKernelType::RHS_Scalar : BinaryKernelType::Strided_RHS_Scalar;
+    if (iter.is_cpu_scalar(scalar_pos)) {
+      scalar = mps::getMPSScalar(otherTensor.item(), otherTensor.scalar_type());
+      otherBuffer = (id<MTLBuffer>)getIMPSAllocator()->allocScalarBufferWithValue(&scalar.value, scalar.size).get();
```

Comment:
This will crash on x86 as it works only on Unified memory (see getMPSGraphTensorFromScalar()).
We can fix later if you think x86 isn't a priority.